ci/builder.go

```diff
@@ -0,0 +1,45 @@
+package main
+
+import (
+	"dagger.io/dagger"
+	"github.com/goreleaser/goreleaser/ci/actions"
+)
+
+// TODO: break up and comment
+func builder(client *dagger.Client, source *dagger.Directory) *dagger.Container {
+	return client.Container().Pipeline("build").
+		From("ubuntu:jammy@sha256:83f0c2a8d6f266d687d55b5cb1cb2201148eb7ac449e4202d9646b9083f1cee0").
+		WithExec([]string{"apt-get", "update"}).
+		WithExec([]string{"apt-get", "install", "-y", "git", "gpg", "nix-bin", "upx-ucl"}).
+		With(actions.SetupDocker).
+		WithExec([]string{"adduser", "-q", "nonroot"}).
+		WithMountedCache("/gomods", client.CacheVolume("gomodcache")).
+		WithExec([]string{"chown", "nonroot", "/gomods"}).
+		WithEnvVariable("GOMODCACHE", "/gomods").
+		WithMountedCache("/gocache", client.CacheVolume("gobuildcache")).
+		WithExec([]string{"chown", "nonroot", "/gocache"}).
+		WithEnvVariable("GOCACHE", "/gocache").
+		WithExec([]string{"mkdir", "/nix"}).
+		WithExec([]string{"chown", "nonroot", "/nix"}).
+		WithMountedDirectory("/src", source).
+		WithExec([]string{"chown", "-R", "nonroot", "/src"}).
+		WithNewFile("/usr/local/sbin/docker", dagger.ContainerWithNewFileOpts{
+			Contents: `#!/bin/sh
+DOCKER_HOST=tcp://localhost:2375 /usr/bin/docker $@`,
```

Comment:
@kpenfound given that @caarlos0 seems to be quite responsive to merge this maybe we can take the opportunity to make this change in goreleaser so this hack is not necessary?

@caarlos0 FWIW we had to do this beacuse `goreleaser` tests seem to overriding env variables when calling docker

Reply:
Agreed. Originally this code ran outside of the goreleaser repo so it was running the tests without modifications. A few workarounds were necessary to accommodate that. In this case, there's a few tests where `localhost` is hardcoded and overriding any `DOCKER_HOST` env

Reply:
hmmm, totally, feel free to make the necessary changes! <3

Reply:
@caarlos0 basically `goreleaser` tests are overriding `go test` envs since they're being set here (https://github.com/goreleaser/goreleaser/blob/7c6201f5f588fa1a8504f5cffa6264646fdee1eb/internal/pipe/docker/docker_test.go#LL992C17-L992C17) and then used here (https://github.com/goreleaser/goreleaser/blob/7c6201f5f588fa1a8504f5cffa6264646fdee1eb/internal/pipe/docker/api.go#L51). 

The change that would help us to remove this hack is if `goreleaser`  appended `ctx.Env` envs to the current existing variables here https://github.com/goreleaser/goreleaser/blob/7c6201f5f588fa1a8504f5cffa6264646fdee1eb/internal/pipe/docker/api.go#L51 instead of overriding them altogether. WDYT?

Reply:
I think it's probably the right thing to do, and what most people would expect to happen.

Honestly kinda surprised no one reported that as a bug yet 🤔

Reply:
> I think it's probably the right thing to do, and what most people would expect to happen.
> 
> Honestly kinda surprised no one reported that as a bug yet thinking

https://github.com/goreleaser/goreleaser/pull/4187

Reply:
merged!